### PR TITLE
Ampers::Aruna

### PR DESCRIPTION
**Pull request by aruna79**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | I first used reservation class to create make_reservation method but decided to create it in the admin class. I also used get_available_rooms as helper method to get all the available rooms in the range so that I can reuse the code in create_block and reserve _available methods
Describe a concept that you gained more clarity on as you worked on this assignment. | I got more clarity on using date class and methods.
Describe a _nominal test_ that you wrote for this assignment. | In "reserve_block_room" it adds the new reserved block to the list of all reservations"
Describe an _edge case test_ that you wrote for this assignment. | "raises error if there are no rooms available in block"
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | Writing pseudocode really helped to plan and structure.I could only write few tests and write the code and then update the remaining tests.Still trying to write all the testsfirst and then write code.

### CODE REVIEW

File: lib/admin_block.rb [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+      rooms_available_range = []
+
+      rooms_available_range = @list_of_rooms
+      (start_date...end_date).each do |date|
+        @all_reservations.each do |res|
+          list =[]
+          if res.start_date <= date && res.end_date > date
+            list << res.room_number
+          end
+          rooms_available_range = rooms_available_range - list
```
Comment (CheezItMan) [inline lib/admin_block.rb:160]: This could be moved to outside the loop.

File: lib/admin_block.rb [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+      return @list_of_rooms - reserved_rooms
+    end
+
+    # To reserve available rooms for a given date range also eliminates rooms that are in reserved in a block
+    def reserve_available_rooms(start_date,end_date)
+
+      rooms_available_range = get_available_rooms(start_date,end_date)
+
+      if rooms_available_range.empty?
+        raise Exception.new("No rooms are available")
```
Comment (CheezItMan) [inline lib/admin_block.rb:76]: Better to `raise StandardError`  It's a bit more specific.

File: lib/admin_block.rb [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        rooms_available_range = rooms_available_range - list_room_block
+        availability = rooms_available_range[0]
+
+        reservation = Hotel::Reservation.new(start_date,end_date,availability)
+        @all_reservations << reservation
+        return @all_reservations
+      end
+    end
+
+    # - As an administrator, I can get the total cost for a given reservation
```
Comment (CheezItMan) [inline lib/admin_block.rb:88]: This is a cost for a date range, not a given reservation.  Small distinct difference.

File: specs/admin_wave1_spec.rb [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+    hotel = Hotel::Admin.new
+    block = hotel.create_block(Date.new(2018,5,5), Date.new(2018,5,9),"one",150)
+
+    block[0].block_id.must_equal "one"
+    block[0].status.must_equal true
+    block[0].room_number.must_be_instance_of Hotel::Room
+    block[0].discount_rate.must_equal 150
+
+  end
+
```
Comment (CheezItMan) [inline specs/admin_wave1_spec.rb:178]: You need more through testing here, including if there aren't enough rooms available in the date range due to reservations.

**Comment by CheezItMan** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Design** |
Demonstrated classes having a single responsibility |  No, all the work is being done in the `Admin` class.  `Reservation` and `Block` have very little to do. 
Demonstrated loose coupling |  It's loosely coupled as the other classes have little to them. 
Methods demonstrate a good use of encapsulation, inputs and outputs | Check
Wave 1 requirements |  You hit the requirements here.
Wave 2 requirements |  Well done, but more testing with reservations is needed.
Wave 3 requirements |  See my notes, more testing is also needed with Reservation 
Summary | Some good work here, you have way too much going on in the `Admin` class and very little in the `Reservation` and `Block` classes.  You've definitely shown some growth, but the design needs work.  Checkout our [reference implementation](https://github.com/Ada-Developers-Academy/textbook-curriculum/blob/master/02-intermediate-ruby/exercises/hotel-revisited.md)
